### DISCUSSION TITLE
Updating UI with multiple tasks

### DISCUSSION THREAD

**Discussion by ashah7** [unoplatform/uno#18518] [Q&A]
Newbie question
How can I update the UI without blocking when carrying out multiple tasks

for example I have two infinite loops setup with different delays to do different tasks
When started I seem to be losing everything inside my TryEnqueue , nothing within this prints to the console, consequently my UI doesn't get updated.  Hope this makes sense, any idea what is wrong here....

So the question is what is the correct way  I update my UI using 2 independent tasks.. As soon as I introduced the  `await anotherdosomething();` under `dispatcher.TryEnqueue` I started getting problems

```

while (true) {
 var dispatcher = Window.Current.DispatcherQueue;
 dispatcher.TryEnqueue(async() =>
 {

      await updatemyinterface();
if (!hasrun) {    <=boolean to ensure this will run only once
      await anotherdosomething();
      hasrun = true
}
});

await Task.Delay(5000);
}

private async Task anotherdosomething() {

while (true) {
    do somethings.....
await Task.Delay(60000);
   }
} 
```

**Comment by MartinZikmund** [comment]
It seems that the first `while (true)` will infinitely enqueue more and more tasks using `TryEnqueue` - is that really what you wanted?

In general `DispatcherQueue.TryEnqueue` is used when you are already on another thread (e.g. doing some work started by `Task.Run`) and need to execute something on the UI thread (e.g. update a progress bar or text on the UI. The code ran by `TryEnqueue` should ideally be quick, as everything that happens on the UI thread (e.g. dispatcher queue bound thread) blocks the UI

**Reply by ashah7** [reply]
yes because this loop is reading some json string locally that if changed need to update my UI depending on the return values.  This need to run for the duration of the app for every 5 seconds
Within this loop there is a another loop that runs say every 40 seconds which also needs to update my UI, but only have the 40 second time lapse.

```
if (!hasrun)
{
    var dispatcher = Window.Current.DispatcherQueue;
    dispatcher.TryEnqueue(async () =>
    {

        Console.WriteLine("running pool for images");
        hasrun = true;
        await Task.Run(anotherdosomething);   //my 40 second nested loop
    });
}
```
I have enclosed the nested loop into a `Task.Run` and that has made some difference, except with the below caveat.

Now what I'm finding when looking at my console is the app seems to get stuck inside the nested loop and returns to the outer loop after the 40s has elapsed.  So if there are any UI updates in the outer loop they will get delayed until it returns from the nested loop.
Fortunately I can tolerate the delayed update.  

Just looking for options really on what's the best way to do something like this

**Reply by MartinZikmund** [reply]
@ashah7 But even in such case there is no need to `TryEnqueue` the whole thing, just the UI manipulation part, for example in pseudocode:

```csharp
void MainLoop()
{
	Task.Run(() => {
		while (true)
		{
			// Work off the UI thread
			dispatcher.TryEnqueue(async () =>
		    {
				// Do something with UI
		    });
			// More work off the UI thread
		}
    }
}
```

Only the update of the UI itself actually needs to run on the UI thread, no need to do the inner loop inside of the `TryEnqueue` action

**Reply by MartinZikmund** [reply]
Also please consider using `DispatcherQueueTimer` which is made to schedule recurring `Tick` on the UI thread (however, the `Tick` even handler actually runs on the UI thread, so do not do any long running work there ideally)
